File: corte_2/ordenamiento/mSort.cpp
#include <iostream>
using namespace std;

// Función que mezcla (merge) dos subarreglos ordenados:
// arr[l..m] y arr[m+1..r]
void merge(int arr[], int l, int m, int r) {
    int n1 = m - l + 1;  // tamaño del subarreglo izquierdo
    int n2 = r - m;      // tamaño del subarreglo derecho

    // arreglos temporales
    int* L = new int[n1];
    int* R = new int[n2];

    // copiar datos a L[] y R[]
    for (int i = 0; i < n1; i++)
        L[i] = arr[l + i];

    for (int j = 0; j < n2; j++)
        R[j] = arr[m + 1 + j];

    // mezclar los subarreglos L[] y R[] en arr[l..r]
    int i = 0;      // índice inicial de L[]
    int j = 0;      // índice inicial de R[]
    int k = l;      // índice inicial de arr[]

    while (i < n1 && j < n2) {
        if (L[i] <= R[j]) {
            arr[k] = L[i];
            i++;
        } else {
            arr[k] = R[j];
            j++;
        }
        k++;
    }

    // copiar los elementos restantes de L[], si hay
    while (i < n1) {
        arr[k] = L[i];
        i++;
        k++;
    }

    // copiar los elementos restantes de R[], si hay
    while (j < n2) {
        arr[k] = R[j];
        j++;
        k++;
    }

    // liberar memoria dinámica
    delete[] L;
    delete[] R;
}

// MergeSort recursivo
void mergeSort(int arr[], int l, int r) {
    if (l < r) {
        int m = l + (r - l) / 2;  // punto medio

        // ordenar primera y segunda mitad
        mergeSort(arr, l, m);
        mergeSort(arr, m + 1, r);

        // mezclar las dos mitades ordenadas
        merge(arr, l, m, r);
    }
}

// Función para imprimir el arreglo
void printArray(int arr[], int size) {
    for (int i = 0; i < size; i++)
        cout << arr[i] << " ";
    cout << endl;
}

// Ejemplo de uso
int main() {
    int arr[] = {10, 7, 8, 9, 1, 5};
    int n = sizeof(arr) / sizeof(arr[0]);

    cout << "Arreglo original:" << endl;
    printArray(arr, n);

    mergeSort(arr, 0, n - 1);

    cout << "Arreglo ordenado:" << endl;
    printArray(arr, n);

    return 0;

File: corte_2/ordenamiento/mSort.py
def merge(arr, l, m, r):
    """
    Mezcla dos sublistas ordenadas: arr[l..m] y arr[m+1..r].
    """
    n1 = m - l + 1  # Tamaño del sub-arreglo izquierdo
    n2 = r - m      # Tamaño del sub-arreglo derecho

    # 1. Crear sublistas temporales (L y R)
    # En Python, se usa slicing para copiar los elementos
    L = arr[l : l + n1]
    R = arr[m + 1 : m + 1 + n2]

    # 2. Inicializar índices
    i = 0     # Índice inicial de L[]
    j = 0     # Índice inicial de R[]
    k = l     # Índice inicial de arr[] (donde comienza la fusión)

    # 3. Mezclar los sub-arreglos de vuelta en arr[l..r]
    while i < n1 and j < n2:
        if L[i] <= R[j]:
            arr[k] = L[i]
            i += 1
        else:
            arr[k] = R[j]
            j += 1
        k += 1

    # 4. Copiar los elementos restantes de L[], si hay
    while i < n1:
        arr[k] = L[i]
        i += 1
        k += 1

    # 5. Copiar los elementos restantes de R[], si hay
    while j < n2:
        arr[k] = R[j]
        j += 1
        k += 1
    
    # Nota: No es necesario 'delete[]' en Python, ya que la recolección
    # de basura se encarga automáticamente de liberar la memoria de L y R.

def merge_sort(arr, l, r):
    """
    Implementación principal recursiva del Merge Sort.
    Ordena el rango arr[l..r].
    """
    if l < r:
        # Encuentra el punto medio para dividir el array
        # Se usa una sintaxis similar a C++ para evitar overflow en números muy grandes,
        # aunque en Python el simple (l + r) // 2 es seguro.
        m = l + (r - l) // 2

        # 1. Ordenar la primera mitad
        merge_sort(arr, l, m)
        # 2. Ordenar la segunda mitad
        merge_sort(arr, m + 1, r)

        # 3. Mezclar las dos mitades
        merge(arr, l, m, r)

def print_array(arr):
    """Función para imprimir la lista."""
    print(arr)

# -----------------
# Programa principal (equivalente a main en C++)
# -----------------
if __name__ == "__main__":
    arr = [10, 7, 8, 9, 1, 5]
    n = len(arr)

    print("Arreglo original:")
    print_array(arr)

    merge_sort(arr, 0, n - 1)

    print("Arreglo ordenado:")
    print_array(arr)
